integration/internal/container/container.go

```diff
@@ -80,3 +84,73 @@ func Inspect(ctx context.Context, t *testing.T, cli client.APIClient, cid string
 
 	return c
 }
+
+type RunResult struct {
+	ContainerID string
+	ExitCode    int
+	Stdout      *bytes.Buffer
+	Stderr      *bytes.Buffer
+}
```

Comment:
Starting to think if we could use `icmd.Result` for these, but it looks like it would need some contributing to upstream because the `outBuffer` and `errBuffer` are not exported.

Could be nice though to use that type for this as well 🤔 

https://github.com/moby/moby/blob/7baab8bd6c0674475f108f4cd1158d9ea30a62cb/vendor/gotest.tools/v3/icmd/command.go#L43-L52


I think some of this code also shows that we need to improve some of the client to not having to write utilities like this (e.g. the demultiplexing). Perhaps something to work on (separately) to move code from the CLI, and/or provide better utilities in the client code 🤔

Reply:
I disagree on all counts.

> Starting to think if we could use `icmd.Result` for these

While there are a lot of similarities, [this is not the result of an `*exec.Cmd` run.](https://en.wikipedia.org/wiki/Circle%E2%80%93ellipse_problem) The gotest.tools module is Apache2 licensed so there's nothing preventing us from copying that code and modifying it to fit the needs of asserting on the result of attaching to a container or running an exec.

> I think some of this code also shows that we need to improve some of the client to not having to write utilities like this (e.g. the demultiplexing).

Demultiplex into what? Buffering the whole demuxed output into an in-memory buffer is rarely the right thing to do in general as the output of an arbitrary command may be unbounded. [We shouldn't provide utilities in the client code which make it easier to do the wrong thing than to do the right thing.](https://blog.codinghorror.com/falling-into-the-pit-of-success/) By only providing `stdcopy` the consumer of the client is forced to choose which `io.Writer` to demux the streams into, giving them the opportunity to consider which choice would best fit their use case.